compiler/transforms/set_memory_space.py

```diff
@@ -188,6 +191,41 @@ def get_cast_op(operand) -> None | memref.MemorySpaceCast:
         rewriter.replace_matched_op(linalg_op)
 
 
+class InitStreamMemorySpace(RewritePattern):
+    """
+    Convert all stream.streaming region ops to only use L1
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: stream.StreamingRegionOp, rewriter: PatternRewriter
+    ):
+        operands_to_memory_cast = tuple(
+            x
+            for x in op.operands
+            if isinstance(x.type, builtin.MemRefType) and x.type.memory_space != L1
+        )
+
+        if not operands_to_memory_cast:
+            return
+
+        # insert memory cast for every value
+        memory_cast_ops: dict[SSAValue, memref.MemorySpaceCast] = {}
+        for operand in operands_to_memory_cast:
+            assert isinstance(memreftype := operand.type, MemRefType)
+            memory_cast_ops[
+                operand
+            ] = memref.MemorySpaceCast.from_type_and_target_space(
+                operand, memreftype, L1
+            )
+        rewriter.insert_op(tuple(memory_cast_ops.values()), InsertPoint.before(op))
+
+        # replace all operands to casted values
+        for i in range(len(op.operands)):
+            if op.operands[i] in operands_to_memory_cast:
+                op.operands[i] = memory_cast_ops[op.operands[i]].dest
+
```

Comment:
Can you rewrite the InitLinalgMemorySpace to be the same like this?
Now these do the pretty much the same, but are written quite differently.
The linalg one reuses previous memory space casts and creates a new op, the way you've written this pass seems more efficient, and to me they look like they can share the exact same code?

Reply:
Okay! I combined them into one pattern!

Reply:
Thank you!